sycl/include/sycl/ext/oneapi/matrix/matrix-jit-use.hpp

```diff
@@ -0,0 +1,660 @@
+};
+
+#define SPV_MATRIX_LAYOUT_TRAITS(LAYOUT, SPV_LAYOUT)                           \
+  template <> struct spv_matrix_layout_traits<LAYOUT> {                        \
+    static constexpr __spv::MatrixLayout value = SPV_LAYOUT;                   \
+  };
+
+SPV_MATRIX_LAYOUT_TRAITS(layout::row_major, __spv::MatrixLayout::RowMajor)
+SPV_MATRIX_LAYOUT_TRAITS(layout::col_major, __spv::MatrixLayout::ColumnMajor)
+SPV_MATRIX_LAYOUT_TRAITS(layout::packed_a, __spv::MatrixLayout::PackedA)
+SPV_MATRIX_LAYOUT_TRAITS(layout::packed_b, __spv::MatrixLayout::PackedB)
+SPV_MATRIX_LAYOUT_TRAITS(layout::unused, __spv::MatrixLayout::Unused)
+
+// unnecessary was introduced for backward compatibility.
+// Once the use implementation is stable, "unnecessary" value will be omitted
+enum class use { a, b, accumulator, unnecessary };
+
+template <use Use> struct spv_matrix_use_traits {
+  static constexpr __spv::MatrixUse value = __spv::MatrixUse::MatrixA;
+};
+
+#define SPV_MATRIX_USE_TRAITS(USE, SPV_USE)                                    \
+  template <> struct spv_matrix_use_traits<USE> {                              \
+    static constexpr __spv::MatrixUse value = SPV_USE;                         \
+  };
+
+SPV_MATRIX_USE_TRAITS(use::a, __spv::MatrixUse::MatrixA)
+SPV_MATRIX_USE_TRAITS(use::b, __spv::MatrixUse::MatrixB)
+SPV_MATRIX_USE_TRAITS(use::accumulator, __spv::MatrixUse::Accumulator)
+SPV_MATRIX_USE_TRAITS(use::unnecessary, __spv::MatrixUse::Unnecessary)
+
+template <typename G> struct spv_scope_traits {};
+template <> struct spv_scope_traits<sycl::sub_group> {
+  constexpr static auto value = __spv::Scope::Subgroup;
+};
+template <int D> struct spv_scope_traits<sycl::group<D>> {
+  constexpr static auto value = __spv::Scope::Workgroup;
+};
+
+template <typename T, size_t NumRows, size_t NumCols, use Use,
+          layout Layout = layout::unused, typename Group = sycl::sub_group>
+class wi_data;
+template <typename T, size_t NumRows, size_t NumCols, use Use,
+          layout Layout = layout::unused, typename Group = sycl::sub_group>
+struct joint_matrix {
+public:
+  __spv::__spirv_JointMatrixINTEL<
+      T, NumRows, NumCols, spv_matrix_layout_traits<Layout>::value,
+      spv_scope_traits<Group>::value, spv_matrix_use_traits<Use>::value> *spvm;
+  joint_matrix(Group sg) {
+#ifndef __SYCL_DEVICE_ONLY__
+    (void)sg;
+    throw runtime_error("joint matrix is not supported on host device.",
+                        PI_ERROR_INVALID_DEVICE);
+#endif // __SYCL_DEVICE_ONLY__
+  }
+
+  inline __SYCL_ALWAYS_INLINE wi_data<T, NumRows, NumCols, Use, Layout, Group>
+  get_wi_data() {
+    return wi_data<T, NumRows, NumCols, Use, Layout, Group>(*this);
+  }
+};
+
+template <typename Group, typename T, size_t NumRows, size_t NumCols, use Use,
+          layout Layout = layout::unused, access::address_space Space>
+inline __SYCL_ALWAYS_INLINE void
+joint_matrix_load(Group sg,
+                  joint_matrix<T, NumRows, NumCols, Use, Layout, Group> &res,
```

Comment:
You can change this to `joint_matrix<T, NumRows, NumCols, Use, layout::unused, Group> &res,`

Then get rid of the layout template parameter as mentioned above.

Reply:
but For CUDA backend, layout must be specified for matrix A and B, cuda users won't set it it as unused for a&b, right?
according to https://github.com/intel/llvm/pull/5835#issuecomment-1224348025

Reply:
That's right for joint_matrix_load; however I think it is better if there is a separate overload for that case of `joint_matrix_load` in the cuda backend like I added here: https://github.com/intel/llvm/blob/a8810552225b4cdb479650011766e9de8f5474b6/sycl/include/sycl/ext/oneapi/matrix/matrix-tensorcore.hpp#L780
Then we have also a version of `joint_matrix_load` that is only for the `accumulator` `matrix::use` here https://github.com/intel/llvm/blob/a8810552225b4cdb479650011766e9de8f5474b6/sycl/include/sycl/ext/oneapi/matrix/matrix-tensorcore.hpp#L752
This immediately prevents the user from using both the template layout parameter and the non-template layout parameter at the same time. This also allows the Intel backend implementation to be simplified in the suggested way.

But for `joint_matrix_store` also in the CUDA backend there is no cases that require the layout as a template parameter (since joint_matrix_store is only functional for matrix::use::accumulator anyway in the CUDA backend.) So also in the Intel backend the layout template parameter is not needed.

It is a little confusing but I hope that makes sense?